---
title: "lab15r"
author: "Rachel Kraft"
date: '2022-03-08'
output: pdf_document
---

> Section 1: Investigating Pertussis by year in the US

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

install.packages("datapasta")

```{r}
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L),
         No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617)
       )
```

```{r message=FALSE}
library(tidyverse)
```

```{r}
ggplot(cdc) +
  aes(x=Year, y=No..Reported.Pertussis.Cases) +
  geom_point() +
  geom_line() +
  labs(y="Number of cases", title="Pertussis cases by year (1922-2019)")
```

> Section 2: Tale of two vaccines

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
ggplot(cdc) +
  aes(x=Year, y=No..Reported.Pertussis.Cases) +
  geom_point() +
  geom_line() +
  labs(y="Number of cases", title="Pertussis cases by year (1922-2019)")+
  geom_vline(xintercept=1946, color="blue", linetype=2) + 
  geom_vline(xintercept=1996, color="red", linetype=2)
```

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

Since the aP vaccine was new, there was possibly more hesitancy in getting the vaccine leading to an increase in observed cases. Also, there could have been more testing that was more accurate, or the bacteria could have mutated and evolved to be more resistant to the vaccines. Another possible reason for the increase in cases is that people who were vaccinated with the old vaccine might have decreased immunity later in life.

> Section 3: Exploring CMI-PB data

Why is this vaccine preventable disease rising? Use jsonlite package to read JSON data, made of key-value pairs where a key is associated with a certain value.

install.packages("jsonlite")

```{r message=FALSE}
library(jsonlite)
```

Let's read the main subject database table from the CMI-PB API

```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
head(subject, 3)
```

> Q4. How may aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$race, subject$biological_sex)
```

> working with dates

```{r}
library(lubridate)
```

we can use the ymd() function to tell lubridate the format of our particular date and then the time_length() function to convert days to years

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

```{r}
subject$age <- today() - ymd(subject$year_of_birth)
time_length(subject$age, "year")
```

```{r message=FALSE}
library(dplyr)
```

```{r}
ap <- subject %>% filter(infancy_vac == "aP")
round( summary( time_length( ap$age, "years" ) ) )
```

```{r}
# wP
wp <- subject %>% filter(infancy_vac == "wP")
round( summary( time_length( wp$age, "years" ) ) )
```

The median age of wP individuals is around 10 years higher than aP individuals 

> Q8. Determine the age of all individuals at time of boost?



> Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```

Read the specimen and ab_titer tables into R and store the data as specimen and titer named data frames.

```{r}
# Complete the API URLs...
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 
```

To know whether a given specimen_id comes from an aP or wP individual we need to link (a.k.a. “join” or merge) our specimen and subject data frames

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details

```{r}
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

> Q12. What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit)
```

The visit 8 specimens are significantly lower than the others, thousands less

> Section 4: examine IgG1 Ab titer levels

Now using our joined/merged/linked abdata dataset filter() for IgG1 isotype and exclude the small number of visit 8 entries

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

FIM2/3, PRN, TT, FHA, DT; These are the antigens that respond to the pertussis bacteria 


We can attempt to examine differences between wP and aP here by setting color and/or facet values of the plot to include infancy_vac status

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  theme_bw()
```

add infancy_vac to the faceting

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

and antigen=="FIM2/3"

```{r}
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

> Q16. What do you notice about these two antigens time course and the FIM2/3 data in particular?

The measles antigens stay relatively constant over time, and the FIM2/3 antigen rises significantly over time, hitting its peak at day 5 and start decreasing again

> Q17. Do you see any clear difference in aP vs. wP responses?

There is not much of a clear difference between aP and wP responses. Possibly the wP antigens rise and drop quicker

> Obtaining CMI-PB RNASeq data

Let's read available RNA-Seq data for the IGHG1 gene into R and investigate the time course of it’s gene expression values

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"
rna <- read_json(url, simplifyVector = TRUE) 
```

To facilitate further analysis we need to “join” the rna expression data with our metadata meta, which is itself a join of sample and specimen data. This will allow us to look at this genes TPM expression values over aP/wP status and at different visits (i.e. times):

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

> Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q19. What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

The expression of this gene is at its highest at visit 4, then drops back down at visit 5 to about the same level it was for visits 1-3

> Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

The antibody titer data shows that antibodies are the highest at visit 5. It makes sense that gene expression is highest at level 4, and by the time of visit 5 the antibodies are made and are the highest then

facet by infancy_vac status

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

There is no obvious wP vs. aP differences here even if we focus in on a particular visit:

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

> Section 6: Working with larger datasets

We will take their “2020 longitudinal RNA-Seq data” file (named 2020LD_rnaseq.csv) from here: https://www.cmi-pb.org/downloads/cmipb_challenge_datasets/2021_cmipb_challenge/

```{r}
rnaseq <- read.csv("2020LD_rnaseq.csv")
head(rnaseq,3)
```

```{r}
dim(rnaseq)
```

how many genes we have reported for each specimen_id

```{r}
n_genes <- table(rnaseq$specimen_id)
head( n_genes , 10)
```

how many specimens?

```{r}
length(n_genes)
```

Check if there are the same number of genes for each specimen

```{r}
all(n_genes[1]==n_genes)
```

let’s convert to wider format with the pivot_wider() function from the tidyr package:

```{r message=FALSE}
library(tidyr)

rna_wide <- rnaseq %>% 
  select(versioned_ensembl_gene_id, specimen_id, tpm) %>%
  pivot_wider(names_from = specimen_id, values_from=tpm)

dim(rna_wide)
```

```{r}
head(rna_wide[,1:7], 3)
```

we can filter this to answer key questions

1. Is RNA-Seq expression levels predictive of Ab titers?
2. What differentiates aP vs wP primed individuals?
3. What about decades after their first immunization?








